subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -102,99 +110,167 @@ def check_payments(self, payments: Iterable[SubscriptionPayment]):
+            # We receive all the elements and check whether we've actually activated them.
+            payment = self._handle_single_receipt_info(request.user, receipt_info)
+            if payment is None:
+                continue
+
+            if latest_payment is None or payment.subscription_end > latest_payment.subscription_end:
+                latest_payment = payment
+
+        if latest_payment is None:
+            logger.warning('No subscription information provided in the payload receipt.')
+            return Response(status=HTTP_400_BAD_REQUEST)
+
+        return Response(SubscriptionPaymentSerializer(latest_payment).data, status=HTTP_200_OK)
 
-        # Find the original transaction, fetch the user, create a new subscription payment.
-        # Note – if we didn't find it, something is really wrong. This notification is only for subsequent payments.
+    def _handle_renewal(self, _user: AbstractBaseUser, notification: AppStoreNotification) -> Response:
+        transaction_info = notification.transaction_info
+
+        # This gets us the very first transaction that happened for this person.
         subscription_payment = SubscriptionPayment.objects.get(
             provider_codename=self.codename,
-            provider_transaction_id=payload.transaction_info.original_transaction_id,
+            provider_transaction_id=transaction_info.original_transaction_id,
         )
 
-        # Currently, we don't support changing of the product ID. Assert here will let us know if anyone did that.
-        # In case the field is not available in metadata, the product ID error will still be raised.
+        # Information about this should've been received earlier via change, but downgrades are to happen
+        # only at the start of new cycle. So we handle them here.
         current_product_id = subscription_payment.plan.metadata.get(self.codename)
-        if current_product_id != payload.transaction_info.product_id:
-            raise ProductIdChangedError(current_product_id, payload.transaction_info.product_id)
+        if current_product_id != transaction_info.product_id:
+            subscription_payment.plan = self._get_plan_for_product_id(transaction_info.product_id)
+            # Create a new subscription.
+            subscription_payment.subscription = None
 
         subscription_payment.pk = None
         # Updating relevant fields.
-        subscription_payment.provider_transaction_id = payload.transaction_info.transaction_id
-        subscription_payment.subscription_start = payload.transaction_info.purchase_date
-        subscription_payment.subscription_end = payload.transaction_info.expires_date
+        subscription_payment.provider_transaction_id = transaction_info.transaction_id
+        subscription_payment.subscription_start = transaction_info.purchase_date
+        subscription_payment.subscription_end = transaction_info.expires_date
         subscription_payment.save()
```

Comment:
```python
subscription_payment.created = None
subscription_payment.updated = None
```
:-O

Reply:
Will that enforce filling with fresh values via a DB trigger?

Reply:
It will via SubscriptionPayment.save(), see method definition